main/src/ca/uwaterloo/flix/language/phase/Lexer.scala

```diff
@@ -43,29 +49,559 @@ object Lexer {
+      case _ if keyword("force") => TokenKind.KeywordForce
+      case _ if keyword("if") => TokenKind.KeywordIf
+      case _ if keyword("import") => TokenKind.KeywordImport
+      case _ if keyword("inline") => TokenKind.KeywordInline
+      case _ if keyword("instance") => TokenKind.KeywordInstance
+      case _ if keyword("into") => TokenKind.KeywordInto
+      case _ if keyword("law") => TokenKind.KeywordLaw
+      case _ if keyword("lawful") => TokenKind.KeywordLawful
+      case _ if keyword("lazy") => TokenKind.KeywordLazy
+      case _ if keyword("let") => TokenKind.KeywordLet
+      case _ if keyword("match") => TokenKind.KeywordMatch
+      case _ if keyword("typematch") => TokenKind.KeywordTypeMatch
+      case _ if keyword("null") => TokenKind.KeywordNull
+      case _ if keyword("override") => TokenKind.KeywordOverride
+      case _ if keyword("pub") => TokenKind.KeywordPub
+      case _ if keyword("ref") => TokenKind.KeywordRef
+      case _ if keyword("region") => TokenKind.KeywordRegion
+      case _ if keyword("sealed") => TokenKind.KeywordSealed
+      case _ if keyword("spawn") => TokenKind.KeywordSpawn
+      case _ if keyword("Static") => TokenKind.KeywordStatic
+      case _ if keyword("true") => TokenKind.KeywordTrue
+      case _ if keyword("type") => TokenKind.KeywordType
+      case _ if keyword("use") => TokenKind.KeywordUse
+      case _ if keyword("where") => TokenKind.KeywordWhere
+      case _ if keyword("with") => TokenKind.KeywordWith
+      case _ if keyword("discard") => TokenKind.KeywordDiscard
+      case _ if keyword("par") => TokenKind.KeywordPar
+      case _ if keyword("yield") => TokenKind.KeywordYield
+      case _ if isMathNameChar(c) => mathName()
+      case _ if isGreekNameChar(c) => greekName()
+
+      // User defined operators
+      case _ if validUserOpTokens.contains(c) => {
+        val p = peek()
+        if (validUserOpTokens.contains(p)) {
+          userDefinedOp()
+        } else if (c == '-' && p.isDigit) {
+          number() // negative numbers
+        } else if (c == '$' && p.isLetter) {
+          builtIn()
+        } else {
+          validUserOpTokens.apply(c)
+        }
+      }
+      case c if c.isLetter => name(c.isUpper)
+      case c if c.isDigit => number()
+      case '\"' => string()
+      case '\'' => char()
+      case '`' => infixFunction()
+      case _ => TokenKind.Err(TokenErrorKind.UnexpectedChar)
+    }
+
+    addToken(kind)
+  }
+
+  // Adds a token by consuming the characters between start and current
+  private def addToken(k: TokenKind)(implicit s: State): Unit = {
+    val t = s.src.data.slice(s.start.offset, s.current.offset).mkString("")
+    s.tokens += Token(k, t, s.start.line, s.start.column)
+    s.start = new Position(s.current.line, s.current.column, s.current.offset)
+  }
+
+  // Checks whether the following substring matches a keyword. Note that *comparison includes current*
+  private def keyword(k: String)(implicit s: State): Boolean = {
```

Comment:
I think this function could make use of some small helper functions.

Reply:
I'm not sure I follow. Which parts would you extract in to helper functions?

Reply:
Perhaps the while loop which computes matches?

Reply:
I have done so but honestly I feel like this reads worse than before. Did I misunderstand you, or do our opinions differ here? 😅 

```scala
  /**
   * Checks whether the following substring matches a keyword. Note that __comparison includes current__.
   */
  private def isMatch(keyword: String)(implicit s: State): Boolean = {
    // Check if the keyword can appear before eof.
    if (s.current.offset + keyword.length > s.src.data.length) {
      return false
    }

    // Check if the next n characters in source matches those of keyword one at a time.
    val start = s.current.offset - 1
    var matches = true
    var offset = 0
    while (matches && offset < keyword.length) {
      if (s.src.data(start + offset) != keyword(offset)) {
        matches = false
      } else {
        offset += 1
      }
    }
    matches
  }

  /**
   * Checks whether the following substring matches a keyword. Note that __comparison includes current__.
   * Also note that this will advance the current position past the keyword if there is a match
   */
  private def isKeyword(keyword: String)(implicit s: State): Boolean = {
    val matches = isMatch(keyword)
    if (matches) {
      for (_ <- 1 until k.length) {
        advance()
      }
    }
    matches
  }
```

Reply:
Hmm okay. I leave it to you then.